### PR TITLE
Add name of C++ source file to generated Python files for ops.

### PR DESCRIPTION
**Pull request by frreiss**
While tracing back through the Python logging APIs to find their implementation, I ran into a dead end at the generated file `gen_logging_ops.py`. This generated file contains no information about where it came from. It took a fair amount of time poking through Bazel build files to track down the rule that generates this file from the output of a program that is linked against an object file that is produced by compiling `logging_ops.cc`.

Other developers don't seem to have gotten that far; for example the reporter https://github.com/tensorflow/tensorflow/issues/8953 , who gave up without finding `logging_ops.cc`.

This PR adds the name of the original C++ source file to the generated Python files for built-in operators. I've modified `python_op_gen_main.cc` to check whether it is running from one of the special executables that the Bazel build creates for these internal operators. If the program is running from inside such a rule, it generates a header comment with the name of the original C++ source file:
```
"""Python wrappers around TensorFlow ops.

This file is MACHINE GENERATED! Do not edit.
Original C++ source file: logging_ops.cc
"""
```
Note the last line, "Original C++ source file...". This should be enough information for someone new to the project to track down the implementation of a built-in operator.

I would have preferred to modify the `REGISTER_OP()` macro to incorporate information about the original source file in the `OpDef` itself, but doing so would have required modifying `op_def.proto`. Hence this approach of checking the executable name in `python_op_gen_main.cc`.

### CODE REVIEW

**Comment by asimshankar** [comment]
Adding @josh11b who has been looking at the python generator lately.

File: tensorflow/python/framework/python_op_gen.cc [hunk 1]
```diff
@@ -710,14 +710,24 @@ string GetPythonOp(const OpDef& op_def, const string& function_name) {
 }
 
 string GetPythonOps(const OpList& ops, const std::vector<string>& hidden_ops,
-                    bool require_shapes) {
+                    const string source_file_name, bool require_shapes) {
```
Comment (josh11b) [inline tensorflow/python/framework/python_op_gen.cc]: In general, please use `const string&` instead of `const string` in function arguments.

**Review by josh11b — CHANGES_REQUESTED** [review]

**Comment by josh11b** [comment]
Thanks for the PR! I think we may eventually need to add the filename to the OpDef itself, but this is a quicker fix.

File: tensorflow/python/framework/python_op_gen.cc (hunk 1 above)
Comment (frreiss) [inline tensorflow/python/framework/python_op_gen.cc]: Sure, no problem. Thanks for the feedback! Updated the PR accordingly.

File: tensorflow/python/framework/python_op_gen_main.cc [hunk 2]
```diff
@@ -78,10 +79,33 @@ Status ParseHiddenOpsCommandLine(const char* arg,
 }
 
 void PrintAllPythonOps(const std::vector<string>& hidden_ops,
+                       const string& source_file_name,
                        bool require_shapes) {
   OpList ops;
   OpRegistry::Global()->Export(false, &ops);
-  PrintPythonOps(ops, hidden_ops, require_shapes);
+  PrintPythonOps(ops, hidden_ops, source_file_name,require_shapes);
```
Comment (josh11b) [inline tensorflow/python/framework/python_op_gen_main.cc]: Space after comma.  I think there is some program to automatically format the code, but I don't have instructions.

**Review by josh11b — CHANGES_REQUESTED** [review]

File: tensorflow/python/framework/python_op_gen_main.cc (hunk 2 above)
Comment (frreiss) [inline tensorflow/python/framework/python_op_gen_main.cc]: Found that program -- `clang-format -style=Google` -- and reformatted the file properly.

**Comment by rmlarsen** [comment]
@josh11b can you take another look, please?

File: tensorflow/python/framework/python_op_gen.cc [hunk 3]
```diff
@@ -713,14 +713,24 @@ string GetPythonOp(const OpDef& op_def, const string& function_name) {
...
   // TODO(josh11b): Mention the library for which wrappers are being generated.
   strings::StrAppend(&result, R"("""Python wrappers around TensorFlow ops.
 
 This file is MACHINE GENERATED! Do not edit.
-"""
+)");
+ 
+  // Mention the original source file so someone tracing back through generated
+  // Python code will know where to look next.
+  if (source_file_name.size() > 0) {
```
Comment (josh11b) [inline tensorflow/python/framework/python_op_gen.cc]: The idiom we use through this code base is to say
if (!source_file_name.empty())

**Review by josh11b — DISMISSED** [review]
This is good with that one change is made.  Thanks!

**Comment by jhseu** [comment]
@frreiss Friendly ping for an update. Thanks!

File: tensorflow/python/framework/python_op_gen.cc (hunk 3 above)
Comment (frreiss) [reply tensorflow/python/framework/python_op_gen.cc]: Sorry, didn't notice this review comment the last time I checked up on the PR. Thanks for the reminder! I've fixed my local copy. Will push the changes once I've merged updates from the trunk and run tests.

**Comment by frreiss** [comment]
Update: Made the remaining requested change from the previous review, then merged with the latest changes from main. On merging, I found that `python_op_gen_main.cc` now uses the routines in `python_eager_op_gen.cc` instead of using `python_op_gen.cc`. So I moved the code that adds the "Original C++ source" notice over to `python_eager_op_gen.cc`. I ran the build, verified that the generated files still get "original c++ source" notices, and ran all the unit tests under `//tensorflow/python/...`.

On a side note, the code in `python_op_gen.cc` and `python_op_gen.h` doesn't appear to be called from anywhere any more. Is that intentional?

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Approving since the changes @josh11b were made.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Pull request merged**